### ISSUE TITLE
feature request: swap w/ receive asset

### ISSUE THREAD

**Issue by dakom** [astroport-fi/astroport-core#206]
**Proposal**

Add a new ExecuteMsg to the pair contracts which enable swaps based on the desired _receive_ amount, instead of _offer_ amount. The returned amount will be the exact receive amount requested, as well as any excess (details below).

**Motivation**

For many swaps, the current API makes perfect sense. You want to give asset A and receive asset B. In this case we aren't so picky about the exact amount of B, within reason (i.e. "slippage tolerance"). Nothing at all to change here, it's great.

However, sometimes this is often not what we want. Instead, we want to receive a precise amount of asset B. The only way to do this currently would be to supply extra asset A, get extra B, and then do _another_ swap of the excess B to A. This is doable, but it's cumbersome and prone to error.

Due to slippage, it's not possible to know the exact amount of A which must be provided in this case. The feature request here is not about doing the impossible - but rather to remove the burden from the user and put it on the contract side, behind this new additional ExecuteMsg. 

Aside for a cleaner API and removing errors in downstream code, it will also reduce gas fees.

**High-level implementation**

_(leaving the exact naming naming conventions out of it, as well as details around fees and taxes)_

1. Client sends an ExecuteMsg with an amount of A and a request to receive an amount of B
2. If there is not enough A to convert into the amount of B, transaction fails (with meaningful error message)
3. If it coincidentally happens to match exactly, then the exact amount of B is returned (and nothing else)
4. If there is excess A, then the exact amount of B is returned *as well as the excess A*

In other words, it's up to the client to query for the current price and provide extra A, but they receive the excess back automatically, as well as the _exact_ amount of B they asked for.

Note that the exact math to work out the amount of A that's required to get B is derivable from x*y=k in the case of the `pair.rs` contract, but it's not a direct multiplier of price iiuc, and the above is probably a bit more complex when we account for fees.

**Comment by dakom** [comment]
Note - a `ReverseSimulation` query may cover this from within a contract, since it returns right away relative to the next execute message, but this is not true when querying and executing from an outside client.

**Comment by stefanionescu** [comment]
Hey, looking at this, we will probably need to update the implementation for every existing pool on Astroport which is already risky. Besides this, the implementation might not be that straightforward or dev friendly.

Will close this PR for now but would be happy to discuss more on the [Astroport Discord](https://discord.gg/QSGKCNtdqw).
